### PR TITLE
BETTER PUSHUP DETECTION

### LINKED ISSUE #3: Improve pushup detection
- Use BlazePose (and adjust the sample rate for the expected 15 FPS performance) to be able to detect hips "through the face"
- Use more points on the body for proper detection
- Calculate elbow angle

### PR DESCRIPTION
**Pull request by ThatXliner**
- Closes #3

### CODE REVIEW

File: src/lib/workouts/pushup.ts [hunk 1]
```diff
@@ -4,40 +4,104 @@ import Workout from '.';
...
+			this.average_left_wrist.y += (left_wrist.y - this.average_left_wrist.y) / this.iterations;
+			if (distance_from_average > 10) {
+				return "Frame rejected; don't move your wrist";
+			}
+		}
+		if (this.average_right_wrist === null) {
+			this.average_right_wrist = right_wrist;
+		} else {
+			const distance_from_average = Math.sqrt(
+				(left_wrist.x - this.average_left_wrist.x) ** 2 +
```
Comment (amaar-mc) [inline src/lib/workouts/pushup.ts]: Why is this calculating the distance from the current **left wrist** to the average **left wrist**? Shouldn't it be right wrist?

File: src/lib/workouts/pushup.ts [hunk 2]
```diff
@@ -4,40 +4,104 @@ import Workout from '.';
...
+					(left_wrist.y - this.average_left_wrist.y) ** 2
+			);
+			this.average_right_wrist.x += (right_wrist.x - this.average_right_wrist.x) / this.iterations;
+			this.average_right_wrist.y += (right_wrist.y - this.average_right_wrist.y) / this.iterations;
+			if (distance_from_average > 10) {
+				return "Frame rejected; don't move your wrist";
+			}
+		}
+		const nose = keypoints.find((keypoint) => keypoint.name === 'nose')!;
```
Comment (amaar-mc) [inline src/lib/workouts/pushup.ts:72]: For clarity move the keypoint up with the rest of the keypoints at the top please

File: src/lib/workouts/pushup.ts [hunk 3]
```diff
@@ -4,40 +4,104 @@ import Workout from '.';
...
+		const left_shoulder = keypoints.find((keypoint) => keypoint.name === 'left_shoulder')!;
+		const right_shoulder = keypoints.find((keypoint) => keypoint.name === 'right_shoulder')!;
+		const left_elbow = keypoints.find((keypoint) => keypoint.name === 'left_elbow')!;
+		const right_elbow = keypoints.find((keypoint) => keypoint.name === 'right_elbow')!;
+		const left_wrist = keypoints.find((keypoint) => keypoint.name === 'left_wrist')!;
+		const right_wrist = keypoints.find((keypoint) => keypoint.name === 'right_wrist')!;
 
-		const left_shoulder_to_elbow = Math.round(
-			Math.sqrt((left_shoulder.x - left_elbow.x) ** 2 + (left_shoulder.y - left_elbow.y) ** 2)
+		// keep a running average
```
Comment (amaar-mc) [inline src/lib/workouts/pushup.ts:42]: What if left_wrist and/or right_wrist are not in frame?

**Review by amaar-mc — CHANGES_REQUESTED** [review]
On my system, when the wrists are not in frame I receive the message "Frame rejected" which should only be executed when there's a location associated for the wrists, and even when they are, as I'm getting into pushup position the counter reports around 5-10 pushups completed already.

File: src/lib/workouts/pushup.ts (hunk 1 above)
Comment (ThatXliner) [reply src/lib/workouts/pushup.ts]: No, it makes sure that wrists stay in around the same place so you can't just cheat by moving your arms closer to your chest (instead you got to move your chest closer to your arms)

File: src/lib/workouts/pushup.ts (hunk 3 above)
Comment (ThatXliner) [reply src/lib/workouts/pushup.ts:42]: They must be in the frame

File: src/lib/workouts/pushup.ts (hunk 2 above)
Comment (ThatXliner) [reply src/lib/workouts/pushup.ts:72]: I'd rather have the code end early when the frame would be rejected either way

**Comment by ThatXliner** [comment]
> as I'm getting into pushup position the counter reports around 5-10 pushups completed already.

Next time, can you send a reproduction video? And can you make sure to tune the values (marked with a comment at the very bottom) to suit you?

**Comment by ThatXliner** [comment]
I added a new UI. I was gonna add a recalibration countdown button but the "stop workout" and "start workout" works for now (this is a prototype, after all).

If the errors only occur "while you're setting up," we will have UI that will give you 5 seconds to set up.

**Review by amaar-mc — APPROVED** [review]
LGTM

**Pull request merged**